File: devices-api-java-mvc/src/test/java/com/lucas/devicesapijavamvc/kafka/DeviceEventPublisherTest.java
package com.lucas.devicesapijavamvc.kafka;

import com.lucas.devicesapijavamvc.domain.Device;
import com.lucas.devicesapijavamvc.domain.DeviceState;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.kafka.core.KafkaTemplate;

import java.time.OffsetDateTime;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

@ExtendWith(MockitoExtension.class)
class DeviceEventPublisherTest {

    private static final String TOPIC = "devices.events";

    @Mock
    KafkaTemplate<String, DeviceEvent> kafkaTemplate;

    @Test
    @DisplayName("publishCreated() should build DeviceEvent and send to Kafka with correct key")
    void publishCreatedShouldSendEvent() {

        Device device = Device.createNew(
                "iPhone 15",
                "Apple",
                DeviceState.AVAILABLE
        );
        OffsetDateTime creationTimeBefore = device.getCreationTime();

        DeviceEventPublisher publisher = new DeviceEventPublisher(kafkaTemplate, TOPIC);
        String origin = "JAVA";

        publisher.publishCreated(device, origin);


        ArgumentCaptor<DeviceEvent> eventCaptor = ArgumentCaptor.forClass(DeviceEvent.class);

        verify(kafkaTemplate).send(
                eq(TOPIC),
                eq(device.getExternalId().toString()),
                eventCaptor.capture()
        );
        verifyNoMoreInteractions(kafkaTemplate);

        DeviceEvent sent = eventCaptor.getValue();
        assertThat(sent).isNotNull();
        assertThat(sent.externalId()).isEqualTo(device.getExternalId());
        assertThat(sent.name()).isEqualTo("iPhone 15");
        assertThat(sent.brand()).isEqualTo("Apple");
        assertThat(sent.state()).isEqualTo(DeviceState.AVAILABLE.name());
        assertThat(sent.origin()).isEqualTo("JAVA");
        assertThat(sent.creationTime()).isEqualTo(creationTimeBefore);
    }
}

File: devices-api-kotlin-reactive/src/test/kotlin/com/lucas/devicesapikotlinreactive/infrastructure/messaging/DeviceEventPublisherTest.kt
package com.lucas.devicesapikotlinreactive.infrastructure.messaging

import com.lucas.devicesapikotlinreactive.domain.model.Device
import com.lucas.devicesapikotlinreactive.domain.model.DeviceState
import io.mockk.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.kafka.core.KafkaTemplate
import reactor.test.StepVerifier
import java.time.OffsetDateTime
import java.util.*

class DeviceEventPublisherTest {

    private val kafkaTemplate = mockk<KafkaTemplate<String, DeviceEvent>>()
    private val kafkaProps = KafkaProperties(
        enabled = true,
        bootstrapServers = "localhost:9092",
        topic = KafkaProperties.Topic(devices = "devices.events.test")
    )

    private val publisher = DeviceEventPublisher(kafkaTemplate, kafkaProps)

    @AfterEach
    fun tearDown() = clearAllMocks()

    @Test
    @DisplayName("publishCreated() sends a DeviceEvent to the configured Kafka topic")
    fun publishCreatedShouldSendEventToKafka() {
        val device = Device(
            id = 1L,
            externalId = UUID.randomUUID(),
            name = "iPhone 15",
            brand = "Apple",
            state = DeviceState.AVAILABLE,
            creationTime = OffsetDateTime.now()
        )

        val topicSlot = slot<String>()
        val keySlot = slot<String>()
        val eventSlot = slot<DeviceEvent>()

        every {
            kafkaTemplate.send(capture(topicSlot), capture(keySlot), capture(eventSlot))
        } returns mockk(relaxed = true)

        val result = publisher.publishCreated(device)

        StepVerifier.create(result)
            .verifyComplete()

        verify(exactly = 1) {
            kafkaTemplate.send(any(), any(), any())
        }

        assertThat(topicSlot.captured).isEqualTo("devices.events.test")
        assertThat(keySlot.captured).isEqualTo(device.externalId.toString())

        val event = eventSlot.captured
        assertThat(event.externalId).isEqualTo(device.externalId)
        assertThat(event.name).isEqualTo(device.name)
        assertThat(event.brand).isEqualTo(device.brand)
        assertThat(event.state).isEqualTo(device.state.name)
        assertThat(event.creationTime).isEqualTo(device.creationTime)
        assertThat(event.origin).isEqualTo("KOTLIN")
    }
}
